### DISCUSSION TITLE
Is it compatible with Solax X-Hybrid SK-TL 5000E?

### DISCUSSION THREAD

**Discussion by GpMissi** [wills106/homeassistant-solax-modbus#25] [Q&A]
Sorry if I don't speak English, I really care about your project and it would be really convenient for me to use it for my Solax. I await a confirmation. Thank you

**Comment by wills106** [comment]
I don't have a full compatibility list yet.

I know the SK-SU5000E works which is a Gen2 X1 Inverter.
What type and brand batteries do you have connected to your inverter?

**Reply by GpMissi** [reply]
Pylontech US2000 Plus

**Reply by wills106** [reply]
You should be ok with the default Gen2 X1 tick box when setting up the component then.

Before altering any of the charge settings just make sure all the values read correctly how you expect them to be.

Is your nominal battery Voltage 48v and max charge / discharge 50A?

**Reply by GpMissi** [reply]
> Is your nominal battery Voltage 48v and max charge / discharge 50A?

Yes 48V but max charge 100A

**Reply by wills106** [reply]
Ahh the SK-TL has the built in BMU (Battery Management Unit) so that's why yours is 100A vs the 50A I have seen with the SK-SU

Everything should work, but if you change the Charge / Discharge from the integration you will be limited to 50A

**Reply by GpMissi** [reply]
Great! I can read all the inverter data, but how can I change the values?

**Reply by wills106** [reply]
I have added them to Known Issues No 8 they don't automatically get added in the Integration page.
Need work on fixing that.

**Reply by GpMissi** [reply]
Ok! I added the hidden entries but I see them as read only.

**Reply by wills106** [reply]
Have you added these?

![image](https://user-images.githubusercontent.com/18155231/154862412-56d6e1d9-a30b-49f7-923a-07416710dfd7.png)

You should get something like

![image](https://user-images.githubusercontent.com/18155231/154862432-20cf34e6-d6b6-4b42-90f8-6a0cca4a5281.png)

Although for some people they have to type in the numbers

![image](https://user-images.githubusercontent.com/18155231/154862475-b71f61f6-4f68-4fb7-87ad-ac5202e904ae.png)

**Reply by wills106** [reply]
It's a bit confusing I know, but if you are using the default configuration to setup SolaX Modbus

![image](https://user-images.githubusercontent.com/18155231/154863263-92e39cb9-0ab7-4036-8b5b-8a744824a487.png)

Either copy or start typing `select.solax_charger_use_mode` into a Lovelace Card, eg Entities Card
You don't need to enable any of the disabled entities.

**Reply by GpMissi** [reply]
![image](https://user-images.githubusercontent.com/34158856/154863488-71d4d638-3121-4a45-ae69-ad7489aaa28a.png)
done, but if I change (for example) the percentage of battery charge fails to do so, it always returns 100%

**Reply by GpMissi** [reply]
which registry should it edit?

**Reply by wills106** [reply]
Do you know what the Battery Minimum Capacity should be set to? Should it be 100% or something else like 10% ?

Does Charge Use Mode or Allow Grid Charge options work?
Did you select the Gen2 X1 option when setting up the component?

**Reply by GpMissi** [reply]
ahahahah!!! Yes! ^_^ 

I set it to 100% to do some tests by reading the Modbus registers with QModMaster

**Reply by GpMissi** [reply]
> Does Charge Use Mode or Allow Grid Charge options work? Did you select the Gen2 X1 option when setting up the component?

I didn't try for fear of doing damage

**Reply by GpMissi** [reply]
If you tell me which registers you are going to edit I can try to read them with QModMaster

Thank you so much for the support!

**Reply by wills106** [reply]
```
NUMBER_TYPES_G2 = [
        ["Battery Minimum Capacity",
            "battery_minimum_capacity",
            0x20,
            "i",
            {
                "min": 0,
                "max": 99,
                "step": 1,
                "unit": PERCENTAGE,
            },
            "battery_capacity_charge"
```

It's set to a Max of 99%, as I didn't expect anyone to set it to 100% Maybe it's stuck at 100% because it's outside of the components parameters. Can you manually set it back to say 15% on the Inverter LCD and see if you can then gain control from HA?

**Reply by wills106** [reply]
Also are you trying to write to registers using this component at the same time as being connected with QModMaster? These SolaX Inverters only allow a single Modbus connection to them I have found when trying to write.

**Reply by GpMissi** [reply]
> Can you manually set it back to say 15% on the Inverter LCD and see if you can then gain control from HA?

I try, I'll be back soon ;-)

**Reply by GpMissi** [reply]
> Also are you trying to write to registers using this component at the same time as being connected with QModMaster? These SolaX Inverters only allow a single Modbus connection to them I have found when trying to write.

I always closed the connection with QModMaster before trying to change from HA

**Comment by GpMissi** [comment]
![image](https://user-images.githubusercontent.com/34158856/154864451-512d956f-f67b-46c4-8a14-27466b1a3aa4.png)

I changed the value from the display

![image](https://user-images.githubusercontent.com/34158856/154864489-4ce9bfee-a470-42dd-98a6-affb225a1cd7.png)

Then I tried to change again

![image](https://user-images.githubusercontent.com/34158856/154864516-39a9c81b-c855-4ed3-a849-0822200a6db7.png)

But back to 20%

![image](https://user-images.githubusercontent.com/34158856/154864530-45e8e850-b5a8-469a-873c-04ad9c8e00fb.png)

**Reply by wills106** [reply]
At least it knows you have changed the value on the Inverter it's self.

Are you setup for the SolaX Cloud or SolaX Portal on this inverter?

**Reply by GpMissi** [reply]
yes

**Reply by wills106** [reply]
Can you change the minimum battery to 25% through the cloud / portal website.
When the integration responds to the 25% change can you then change the value back to 20% within the integration?

**Reply by GpMissi** [reply]
> Can you change the minimum battery to 25% through the cloud / portal. When the integration responds to the 25% change can you then change the value back to 20% within the integration?

I use a portal from which I cannot make changes

![image](https://user-images.githubusercontent.com/34158856/154864945-028c8861-6a8a-4090-91dc-18607ca16263.png)

**Comment by GpMissi** [comment]
One question: is it not possible to change the two discharge ranges of the battery as well?

**Reply by wills106** [reply]
Do you have an example?

I am starting to wonder if your SK-TL has yet another different set of Modbus commands?
It seems they keep 90 - 95% the same, but do a subtle change each time. So far I have docs for SK-SU (Gen 2 non BMU) Gen3 and Gen 4

**Reply by GpMissi** [reply]
![image](https://user-images.githubusercontent.com/34158856/154865621-d3fd312d-4693-41bb-b0e7-fff617901803.png)
![image](https://user-images.githubusercontent.com/34158856/154865626-a02c39ab-f42b-40ea-a8f0-b4737ad2460b.png)
![image](https://user-images.githubusercontent.com/34158856/154865644-c0aabdb7-0aec-43b2-bc51-1b32179b748f.png)
![image](https://user-images.githubusercontent.com/34158856/154865651-b5186d2e-84f8-4f09-b8c8-095fc1146484.png)
![image](https://user-images.githubusercontent.com/34158856/154865667-1f968ed4-8ee4-4ab0-b5a0-3c192c174594.png)
![image](https://user-images.githubusercontent.com/34158856/154865674-3fb9a5f4-7ef0-4153-aab1-009814a08365.png)
![image](https://user-images.githubusercontent.com/34158856/154865704-c5a8e0a8-8fcf-4dc4-bcba-c7a3e9745df1.png)
![image](https://user-images.githubusercontent.com/34158856/154865690-f7c7feda-85f7-4901-b132-dd4581663296.png)
![image](https://user-images.githubusercontent.com/34158856/154865735-c7bc12db-a6f9-40f4-b5b7-6005540a03d9.png)
![image](https://user-images.githubusercontent.com/34158856/154865746-eedc6b90-27bc-4a82-a391-0fc5ad88ae97.png)
![image](https://user-images.githubusercontent.com/34158856/154865748-7e0d22ff-9509-4d02-b00a-b91ddccb4d52.png)
![image](https://user-images.githubusercontent.com/34158856/154865759-0fba7318-b73d-4402-986f-8b74eb2d8474.png)

**Reply by GpMissi** [reply]
Before that I had a SU5000E and the menu was the same

**Reply by GpMissi** [reply]
![image](https://user-images.githubusercontent.com/34158856/154868314-3a338a0f-c745-4285-92d8-3ac55e1d1181.png)
red = time ss-mm-hh (from 85 to 87)
Blue = date dd-mm-yy (from 88 to 8A)
green = Charge start time 1 - charge end time 1 / charge start time 2 - charge end time 2 
yellow = Discharge start time 1 - discharge end time 1 / discharge start time 2 - discharge end time 2

**Reply by wills106** [reply]
Reading is already supported.
Red and Blue are the RTC values, but by default the RTC is disabled

You can read times:
![image](https://user-images.githubusercontent.com/18155231/154898727-5ba13a62-67c6-40c5-a081-99783de43ef7.png)

The problem though is that I have not worked out an easy way to program them within HA.
The first problem is that for built in Integrations (custom_components) there is no equivalent to input.date_time so I would either need a separate number entry for hours and mins for each time block. So a total of 8 number.solax_ or look into how I monitor a user created input.date_time
Which would either need the user to follow an exact naming convention, or for the option to put in the various input.date_time into the initial Integration Configuration page. So it knows what to monitor.

Then to complicate things more SolaX decided that while you read hours as a single 16bit entry and minutes as another 16bit entry, you program both the hours and minutes for a timeslot as a single 16bit number. Using hi 8bit for mins and the low 8bit as the hours.

So if you want to program the Start Time of Slot 1 you need to do the following for 09:10 am

Hbyte = 10
Lbyte = 9
value = Hbyte << 8 | Lbyte;
print(value)

Which gives you the number of 2569 which is what you program to the corresponding register.

**Reply by wills106** [reply]
Have you been able to try out 0.4.4 yet? It has basic control for the charge periods.
If the values appear to be correct I can push out 0.4.5 with the rest of them.

**Comment by GpMissi** [comment]
Any idea on how I can write the data on my firmware? What data can I provide you to try to investigate the problem and find a solution? Thank you

**Reply by wills106** [reply]
I added the following into 0.4.4 / 0.4.5

button.solax_battery_awaken
number.solax_battery_charge_max_current
number.solax_battery_discharge_max_current
select.solax_charger_end_time_1
select.solax_charger_end_time_2
select.solax_charger_start_time_1
select.solax_charger_start_time_2
select.solax_charger_use_mode

Do they not work for you? They give you control over when / how the battery charges. (Charge times are 15 minute resolution)

**Reply by GpMissi** [reply]
I have two problems:
1) I also have two discharge ranges (see above)
2) I can't change any of the editable 

![image](https://user-images.githubusercontent.com/34158856/159303575-e4559996-de3b-48a1-a835-35a494b3d070.png)

**Reply by wills106** [reply]
You need to manually add in the number.x and select.x into your Lovelace dashboard.
If you search on your "Developer Tools" page you will find the exact naming.

**Reply by GpMissi** [reply]
![image](https://user-images.githubusercontent.com/34158856/159310590-67490424-9842-41b6-9b93-f2659d6bf77a.png)

**Reply by wills106** [reply]
Yea you need to manually create a Lovelace card for them if you want to see them.

```
entities:
  - entity: select.solax_a66f4ed3_charger_use_mode
  - entity: select.solax_a66f4ed3_allow_grid_charge
  - entity: number.solax_a66f4ed3_battery_charge_max_current
  - entity: number.solax_a66f4ed3_battery_discharge_max_current
  - entity: number.solax_a66f4ed3_battery_minimum_capacity
  - entity: number.solax_a66f4ed3_forcetime_period_1_max_capacity
  - entity: number.solax_a66f4ed3_forcetime_period_2_max_capacity
  - entity: select.solax_a66f4ed3_charger_start_time_1
  - entity: select.solax_a66f4ed3_charger_end_time_1
  - entity: select.solax_a66f4ed3_charger_start_time_2
  - entity: select.solax_a66f4ed3_charger_end_time_2
  - entity: button.solax_a66f4ed3_battery_awaken
show_header_toggle: false
title: SolaX Battery
type: entities
```

If you have times set other than in 15min blocks they won't show correct.
ie 01:05 won't show, but 01:15 will do. Or 01:00, 01:30, 01:45

**Reply by GpMissi** [reply]
Forgive me, but I don't speak English and I use google translator. I am explaining myself badly. I see the values ​​for the charge but even if I try to modify them they remain as they are to those set by the inverter panel. On the other hand, the two hourly ranges relating to the discharge are missing

![image](https://user-images.githubusercontent.com/34158856/159312183-ee37edbf-b1ce-4d8d-a347-e0c2071ec4a0.png)

**Reply by wills106** [reply]
Sorry I missed the bit about Discharge.

What exactly happens when you change the Discharge times?
When we tried the registers out on a SK-SU5000E it didn't force discharge the batteries into the grid.
So we couldn't work out eactly what it's meant to do?

**Reply by GpMissi** [reply]
the discharge time allows you to use the battery only from the start time to the end time. This is handy to avoid using battery even when electricity is cheaper.

**Reply by wills106** [reply]
Ok that makes sense. We were expecting it to empty the battery into the Grid, so we couldn't work out exactly what it was meant to do.

I'll add them back in.

**Reply by GpMissi** [reply]
My biggest problem remains that I can't change these values ​​on my inverter. Do you think there are tests that I can do to understand if I can change the parameters from HA?

**Reply by wills106** [reply]
We have previously been able to change those "Discharge" times. On both a SK-SU5000E and my Gen3 X1-Hybrid-5.0-D-E within this Integration. It wasn't released publicy as we didn't think it actually did anything usefull.

But like I say I can add them back in.

**Reply by wills106** [reply]
If you can't change those times from your Inverter, how did they get set in the first place?

**Reply by GpMissi** [reply]
My most expensive rate goes from 7 am to 11 pm. The most convenient thing for me is to charge until 6:59 am and from 7 am the discharge starts which behaves like the Self Use mode. During the day, therefore, the battery charges and discharges according to consumption and the production of the panels. At 11 pm what is left in the batteries I keep it to be able to exploit it in the morning from 7 onwards.

**Reply by GpMissi** [reply]
I can change the settings from the inverter. I can't change any parameters from HA.

**Reply by wills106** [reply]
What mode do these discharge times work in? Only "Force Time" or all modes?

**Reply by GpMissi** [reply]
Only "Force Time"

**Reply by wills106** [reply]
I have added them to [0.4.10](https://github.com/wills106/homsassistant-solax-modbus/releases/tag/0.4.10)

**Comment by MJP-76** [comment]
When I put them in with the “experimental” settings I could change both charge and discharge settings

SK-SU5000E

Sent from my iPhone

On 21 Mar 2022, at 17:43, Gianpaolo ***@***.***> wrote:

﻿

I can change the settings from the inverter. I can't change any parameters from HA.

**Reply by GpMissi** [reply]
What does this "experimental" mode consist of?

**Reply by wills106** [reply]
I sent @matthewjporter a set a test files to see what the Discharge did on his Inverter.
I never released them publicly.

I can add them back into a public release soon.

**Reply by GpMissi** [reply]
excellent

**Comment by GpMissi** [comment]
I also read this error, I don't know what it refers to:

Logger: homeassistant.components.sensor.recorder
Source: components/sensor/recorder.py:311
Integration: Sensore (documentation, issues)
First occurred: 21 marzo 2022, 11:25:10 (1 occurrences)
Last logged: 21 marzo 2022, 11:25:10

Entity sensor.solax_a66f4ed3_battery_input_energy_today from integration solax_modbus has state class total_increasing, but its state is not strictly increasing. Triggered by state 8.9 (9.8) with last_updated set to 2022-03-21T10:22:14.052704+00:00. Please report it to the custom component author.

**Reply by wills106** [reply]
battery_input_energy_today
battery_output_energy_today

Are not meant to be there for the Gen2 Inverter. I might have forgot to remove them for the Gen2 or added them back in as a mistake at some point.

**Comment by GpMissi** [comment]
Among the parameters of the "Work mode" there is a "REMOTE". Can you tell me how it works? Thank you

**Reply by wills106** [reply]
@matthewjporter Is Remote mode on the Gen2 the same as Backup on Gen3? Or do I need to split them up.
On the Gen3 it charges the battery up and doesn't discharge the battery till you allow it to. This would be handy to save the battery charge to a time when the power supply might be intermittent if you had the EPS setup. Or you want to save the battery for a high peak import cost.

**Reply by GpMissi** [reply]
I have no idea :-(

**Reply by GpMissi** [reply]
I have to find a way to drive the use of the battery from HA. I find nothing that allows me to do this and I don't know which way to look

**Reply by MJP-76** [reply]
> 

I only have 3 options. "Self Use" "Force Time Use" & "Remote Control"

It seems the following is how the Gen2 works
"Self Use" Solar & Battery is used as needed and additional Solar will go into the battery automatically. I use this most of the day
"Force Time Use" Switches to charge batteries from the Grid. I use overnight on Octopus Agile Go cheap rate or on Agile cheap rate
"Remote Control" I have no idea what this actually means but when I put the inverters into "Backup Mode" in HA they stop using the battery so would assume they are going into "EPS" and I do use this

My guess is that "Remote Mode" was going to be included in the solax-portal site but never implemented on my inverter/portal combo but was on later inverters and the solax-cloud site

Does that make sense

**Reply by wills106** [reply]
This is the extract from the Gen3 manual for Backup Mode. Which makes sense for what you are seeing with it not discharging your battery on the Gen2 in "Remote Control" mode.

```
This mode applies the area that has frequent power outages. And this mode ensures the battery will has enough energy to supply when the grid is off.

In this mode battery will be charging forcibly in the setting time and will never be discharged when the grid is on, and it also allows to choose whether charge from the grid or not.
```

While it might be designed for EPS mode. It can be handy to store power in the battery in Remote / Backup mode and then switch back to Self Use, for example during a period of time when the energy costs are high.

**Reply by MJP-76** [reply]
Exactly how I see it. Used to store energy for when needed or stop the discharge when not needed

I put into this mode when my electric heaters are on so as not to drain the battery on heating but rather when needed to run the base load etc

**Comment by GpMissi** [comment]
WOW! The key to unlock the inverter works and now I can change the values ​​from HA! Great

**Comment by GpMissi** [comment]
Good morning. Now that I can change the values ​​of my inverter I would like to ask if it is possible to add the following features:

1) The Gen2 "SK-TL" can go up to 100A, is it possible to extend this value up to 100?
![image](https://user-images.githubusercontent.com/34158856/160067639-eb0a1e22-5bc0-4e9a-bab0-11b8bf89b3d9.png)

2) Would it be possible to add a button or a function to synchronize the inverter time? The inverter clock is often unreliable and needs to be updated often.

3) The start and end times of charge / discharge should be freely modifiable with the precision of the single minute

Could I have some examples of automation in HA to understand how it can change the values ​​when an event occurs?

Thanks for the wonderful work.
If I am able to create HA automations to change the values ​​automatically, I will be happy to send my contribution.

**Reply by wills106** [reply]
> * The Gen2 "SK-TL" can go up to 100A, is it possible to extend this value up to 100?
>   ![image](https://user-images.githubusercontent.com/34158856/160067639-eb0a1e22-5bc0-4e9a-bab0-11b8bf89b3d9.png)

For now you can change:

https://github.com/wills106/homsassistant-solax-modbus/blob/466596ab7a267d23751444bd23a7b995723ad9a2/custom_components/solax_modbus/const.py#L88

https://github.com/wills106/homsassistant-solax-modbus/blob/466596ab7a267d23751444bd23a7b995723ad9a2/custom_components/solax_modbus/const.py#L99

Both to 100. This is where I need to start building in a check based on the first 4 or 5 digits of the Serial Number, otherwise we will end up with 50 tick boxes on initial setup for subtle differences.

> * Would it be possible to add a button or a function to synchronize the inverter time? The inverter clock is often unreliable and needs to be updated often.

Writing to multiple registers at once is not something I have looked into yet.

> * The start and end times of charge / discharge should be freely modifiable with the precision of the single minute

#37 Mentions Dynamic times, they are all hard encoded at the moment as they use a high byte and low byte to a single 16bit register.

> * Could I have some examples of automation in HA to understand how it can change the values ​​when an event occurs?

#38 Is the only real example I have so far. I don't what what others have done?

**Reply by GpMissi** [reply]
Good. I had already made the first modification proposed.

**Reply by GpMissi** [reply]
4) my Solax firmware has only 3 "Work mode" values:
- Self Use Mode
- Force Time Use
- Remote

**Reply by wills106** [reply]
Not got round to removing the 4th option for Work mode on the Gen2 yet and also from what I understand the "Remote" option doesn't do anything either.

**Reply by GpMissi** [reply]
Yes, I still haven't been able to figure out what the "remote" mode does

**Reply by wills106** [reply]
@GpMissi I don't know if you have seen [0.4.12b](https://github.com/wills106/homsassistant-solax-modbus/releases/tag/0.4.12b) & [0.4.12b-1](https://github.com/wills106/homsassistant-solax-modbus/releases/tag/0.4.12b-1) Yet or not. If you try them remember you need to change "battery_charge_max_current" & "battery_discharge_max_current" to 100 again.

**Reply by wills106** [reply]
[0.4.13](https://github.com/wills106/homsassistant-solax-modbus/releases/tag/0.4.13) Contains the Battery Energy changes for the Dashboard. But you still need to manually change "battery_charge_max_current" & "battery_discharge_max_current" to 100 again.

**Reply by wills106** [reply]
@GpMissi have you tested any of the pre-release 0.5.0b versions?

Does your Inverter get detected ok and your charge / discharge get set to 100A correctly?
